File: src/typecheck.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::fmt::Display;
use std::hash::Hash;
use std::sync::atomic::{AtomicU32, Ordering};

static VAR_ID: AtomicU32 = AtomicU32::new(0);

use crate::ast_util::{Printer, Symbol};
use crate::{ast::*, union_find::UnionFind};

#[derive(Clone)]
pub struct Constraint {
    /// The type on the left.
    type_l: Type,
    /// The type on the right.
    type_r: Type,
    /// A string that represents the left type for debug purpose.
    expr_l: String,
    /// A string that represents the right type for debug purpose.
    expr_r: String,
}

macro_rules! flat {
    ($vec:expr) => {
        $vec.into_iter().flatten().collect()
    };
}

fn vec_diff<T>(vec1: Vec<T>, vec2: Vec<T>) -> Vec<T>
where
    T: Hash + Eq + Clone,
{
    let vec1_set = HashSet::<_>::from_iter(vec1);
    let vec2_set = HashSet::<_>::from_iter(vec2);
    vec1_set.difference(&vec2_set).cloned().collect()
}

trait GetVars {
    fn get_all_vars(&self) -> Vec<Variable>;
    fn get_scoped_vars(&self) -> Vec<Variable>;
    fn get_free_vars(&self) -> Vec<Variable> {
        vec_diff(self.get_all_vars(), self.get_scoped_vars())
    }
}

impl<T> GetVars for Vec<T>
where
    T: GetVars,
{
    fn get_all_vars(&self) -> Vec<Variable> {
        self.iter().flat_map(|x| x.get_all_vars()).collect()
    }
    fn get_scoped_vars(&self) -> Vec<Variable> {
        self.iter().flat_map(|x| x.get_scoped_vars()).collect()
    }
}

impl GetVars for Type {
    fn get_scoped_vars(&self) -> Vec<Variable> {
        match self {
            Type::Bool | Type::Num | Type::Unit | Type::Var(_) => vec![],
            Type::Fn { arg, ret } => arg
                .get_scoped_vars()
                .into_iter()
                .chain(ret.get_scoped_vars())
                .collect(),
            Type::Product { left, right } | Type::Sum { left, right } => left
                .get_scoped_vars()
                .into_iter()
                .chain(right.get_scoped_vars())
                .collect(),
            Type::Forall { a, tau } => std::iter::once(a.clone())
                .chain(tau.get_scoped_vars())
                .collect(),
            _ => todo!(),
        }
    }

    fn get_all_vars(&self) -> Vec<Variable> {
        match self {
            Type::Bool | Type::Num | Type::Unit => vec![],
            Type::Var(x) => vec![x.clone()],
            Type::Fn { arg, ret } => arg
                .get_all_vars()
                .into_iter()
                .chain(ret.get_all_vars())
                .collect(),
            Type::Product { left, right } | Type::Sum { left, right } => left
                .get_all_vars()
                .into_iter()
                .chain(right.get_all_vars())
                .collect(),
            Type::Forall { a, tau } => std::iter::once(a.clone())
                .chain(tau.get_all_vars())
                .collect(),
            _ => todo!(),
        }
    }
}

impl Type {
    fn instantiate(self) -> Self {
        match self {
            Type::Bool | Type::Num | Type::Unit | Type::Var(_) => self,
            Type::Forall { a, tau } => tau.instantiate().substitute(a, fresh_type_var()),
            Type::Sum { left, right } => Type::Sum {
                left: left.instantiate().into(),
                right: right.instantiate().into(),
            },
            Type::Product { left, right } => Type::Product {
                left: left.instantiate().into(),
                right: right.instantiate().into(),
            },
            Type::Fn { arg, ret } => Type::Fn {
                arg: arg.instantiate().into(),
                ret: ret.instantiate().into(),
            },
            _ => todo!(),
        }
    }

    fn generalize(self, ctx: &Vec<Type>, constraints: Vec<Constraint>) -> Result<Type, String> {
        let (uf, map) = unification(constraints)?;
        let mut tau_x = get_type(self, uf, map);
        for a in vec_diff(tau_x.get_free_vars(), ctx.get_free_vars()) {
            tau_x = tau_x.add_one_quantifier(a);
        }
        Ok(tau_x)
    }

    fn add_one_quantifier(self, a: Variable) -> Self {
        Type::Forall {
            a,
            tau: self.into(),
        }
    }
}

impl Expr {
    pub fn get_constraints(&self, ctx: &mut Vec<Type>) -> Result<(Type, Vec<Constraint>), String> {
        match self {
            // 1. arithmetic
            Expr::Num(_) => Ok((Type::Num, vec![])),
            Expr::Addop { left, right, .. } | Expr::Mulop { left, right, .. } => {
                let (tau_left, c_left) = left.get_constraints(ctx)?;
                let (tau_right, c_right) = right.get_constraints(ctx)?;
                let constraints = flat!(vec![
                    c_left,
                    c_right,
                    vec![
                        Constraint {
                            type_l: tau_left,
                            type_r: Type::Num,
                            expr_l: left.to_string(),
                            expr_r: "Num".to_string()
                        },
                        Constraint {
                            type_l: tau_right,
                            type_r: Type::Num,
                            expr_l: right.to_string(),
                            expr_r: "Num".to_string()
                        },
                    ]
                ]);
                Ok((Type::Num, constraints))
            }
            // 2. conditionals
            Expr::True | Expr::False => Ok((Type::Bool, vec![])),
            Expr::Relop { left, right, .. } => {
                let (tau_left, c_left) = left.get_constraints(ctx)?;
                let (tau_right, c_right) = right.get_constraints(ctx)?;
                let constraints = flat!(vec![
                    c_left,
                    c_right,
                    vec![
                        Constraint {
                            type_l: tau_left,
                            type_r: Type::Num,
                            expr_l: left.to_string(),
                            expr_r: "Num".to_string()
                        },
                        Constraint {
                            type_l: tau_right,
                            type_r: Type::Num,
                            expr_l: right.to_string(),
                            expr_r: "Num".to_string()
                        },
                    ]
                ]);
                Ok((Type::Bool, constraints))
            }
            Expr::If { cond, then_, else_ } => {
                let (tau_cond, c_cond) = cond.get_constraints(ctx)?;
                let (tau_then, c_then) = then_.get_constraints(ctx)?;
                let (tau_else, c_else) = else_.get_constraints(ctx)?;
                let constraints = flat!(vec![
                    c_cond,
                    c_then,
                    c_else,
                    vec![
                        Constraint {
                            type_l: tau_cond,
                            type_r: Type::Bool,
                            expr_l: cond.to_string(),
                            expr_r: "Bool".to_string()
                        },
                        Constraint {
                            type_l: tau_then.clone(),
                            type_r: tau_else,
                            expr_l: then_.to_string(),
                            expr_r: else_.to_string()
                        },
                    ]
                ]);
                Ok((tau_then, constraints))
            }
            Expr::And { left, right } | Expr::Or { left, right } => {
                let (tau_left, c_left) = left.get_constraints(ctx)?;
                let (tau_right, c_right) = right.get_constraints(ctx)?;
                let constraints = flat!(vec![
                    c_left,
                    c_right,
                    vec![
                        Constraint {
                            type_l: tau_left,
                            type_r: Type::Bool,
                            expr_l: left.to_string(),
                            expr_r: "Bool".to_string()
                        },
                        Constraint {
                            type_l: tau_right,
                            type_r: Type::Bool,
                            expr_l: right.to_string(),
                            expr_r: "Bool".to_string()
                        },
                    ]
                ]);
                Ok((Type::Bool, constraints))
            }
            // 3. functions
            Expr::Var(x) => Err(format!("Free variable: {x}")),
            Expr::DeBruijn(depth) => Ok((
                ctx.iter()
                    .rev()
                    .skip(*depth)
                    .next()
                    .unwrap()
                    .clone()
                    .instantiate(),
                vec![],
            )),
            Expr::Lam { x: _, e } => {
                let tau = fresh_type_var();
                ctx.push(tau);
                let (tau_ret, c_ret) = e.get_constraints(ctx)?;
                let tau = ctx.pop().unwrap();
                Ok((
                    Type::Fn {
                        arg: Box::new(tau),
                        ret: Box::new(tau_ret),
                    },
                    c_ret,
                ))
            }
            Expr::App { lam, arg } => {
                let (tau_lam, c_lam) = lam.get_constraints(ctx)?;
                let (tau_arg, c_arg) = arg.get_constraints(ctx)?;
                let tau_ret = fresh_type_var();
                let constraints = flat!(vec![
                    c_lam,
                    c_arg,
                    vec![Constraint {
                        type_l: tau_lam,
                        expr_r: format!("{} → {}", tau_arg, tau_ret),
                        type_r: Type::Fn {
                            arg: Box::new(tau_arg),
                            ret: Box::new(tau_ret.clone()),
                        },
                        expr_l: lam.to_string()
                    }]
                ]);
                Ok((tau_ret, constraints))
            }
            // 4. product types
            Expr::Pair { left, right } => {
                let (tau_l, c_l) = left.get_constraints(ctx)?;
                let (tau_r, c_r) = right.get_constraints(ctx)?;
                let constraints = flat!(vec![c_l, c_r]);
                Ok((
                    Type::Product {
                        left: Box::new(tau_l),
                        right: Box::new(tau_r),
                    },
                    constraints,
                ))
            }
            Expr::Project { e, d } => {
                let (tau_e, c_e) = e.get_constraints(ctx)?;
                let tau_l = fresh_type_var();
                let tau_r = fresh_type_var();
                let constraints = flat!(vec![
                    c_e,
                    vec![Constraint {
                        type_l: tau_e,
                        expr_r: format!("({} * {})", tau_l, tau_r),
                        type_r: Type::Product {
                            left: Box::new(tau_l.clone()),
                            right: Box::new(tau_r.clone()),
                        },
                        expr_l: e.to_string()
                    }]
                ]);
                let tau_ret = match d {
                    Direction::Left => tau_l,
                    Direction::Right => tau_r,
                };
                Ok((tau_ret, constraints))
            }
            Expr::Unit => Ok((Type::Unit, vec![])),
            // 5. sum types
            Expr::Inject { e, d } => {
                let (tau_e, c_e) = e.get_constraints(ctx)?;
                let tau_other = fresh_type_var();
                let tau_full = match d {
                    Direction::Left => Type::Sum {
                        left: Box::new(tau_e.clone()),
                        right: Box::new(tau_other.clone()),
                    },
                    Direction::Right => Type::Sum {
                        left: Box::new(tau_other.clone()),
                        right: Box::new(tau_e.clone()),
                    },
                };
                Ok((tau_full, c_e))
            }
            Expr::Case {
                e, eleft, eright, ..
            } => {
                let (tau_sum, c_sum) = e.get_constraints(ctx)?;
                let tau_l = fresh_type_var();
                let tau_r = fresh_type_var();

                ctx.push(tau_l);
                let (tau_l_after, c_l) = eleft.get_constraints(ctx)?;
                let tau_l = ctx.pop().unwrap();
                ctx.push(tau_r);
                let (tau_r_after, c_r) = eright.get_constraints(ctx)?;
                let tau_r = ctx.pop().unwrap();

                let constraints = flat!(vec![
                    c_sum,
                    c_l,
                    c_r,
                    vec![
                        Constraint {
                            type_l: tau_sum,
                            expr_r: format!("({}+{})", tau_l, tau_r),
                            type_r: Type::Sum {
                                left: Box::new(tau_l.clone()),
                                right: Box::new(tau_r.clone()),
                            },
                            expr_l: e.to_string()
                        },
                        Constraint {
                            type_l: tau_l_after.clone(),
                            type_r: tau_r_after,
                            expr_l: eleft.to_string(),
                            expr_r: eright.to_string()
                        },
                    ]
                ]);
                Ok((tau_l_after, constraints))
            }
            // 6. fixpoints
            Expr::Fix { x, e } => {
                let tau_x = fresh_type_var();
                ctx.push(tau_x);
                let (tau_e, c_e) = e.get_constraints(ctx)?;
                let tau_x = ctx.pop().unwrap();
                let constraints = flat!(vec![
                    c_e,
                    vec![Constraint {
                        type_l: tau_x.clone(),
                        type_r: tau_e,
                        expr_l: x.to_string(),
                        expr_r: e.to_string()
                    }]
                ]);
                Ok((tau_x, constraints))
            }
            // 7. polymorphism
            Expr::Let { e_x, e_in, .. } => {
                let (tau_x, c_x) = e_x.get_constraints(ctx)?;
                ctx.push(tau_x.generalize(ctx, c_x.clone())?);
                let (tau_in, c_in) = e_in.get_constraints(ctx)?;
                let _ = ctx.pop().unwrap();
                Ok((tau_in, flat!(vec![c_in, c_x])))
            }
            _ => todo!(),
        }
    }

    fn type_check(self) -> Result<Type, String> {
        let mut ctx = vec![];
        let (cur_type, constraints) = self.get_constraints(&mut ctx)?;
        assert!(ctx.is_empty());
        let (uf, map) = unification(constraints)?;
        Ok(get_type(cur_type, uf, map))
    }
}

impl Display for Constraint {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}: {} = {}: {}",
            self.expr_l, self.type_l, self.expr_r, self.type_r
        )
    }
}

pub fn type_check(ast: &Expr) -> Result<Type, String> {
    let ty = ast.clone().to_debruijn().type_check()?;
    assert!(ty.get_free_vars().is_empty());
    Ok(ty)
}

fn fresh_type_var() -> Type {
    let cur_id = VAR_ID.fetch_add(1, Ordering::Relaxed);
    Type::Var(Variable(format!("type_{cur_id}")))
}

fn unification(
    constraints: Vec<Constraint>,
) -> Result<(UnionFind<Variable>, HashMap<Variable, Type>), String> {
    let variables: Vec<Variable> = constraints
        .iter()
        .flat_map(|c| vec![&c.type_l, &c.type_r])
        .flat_map(|t| t.get_all_vars())
        .collect();
    let mut uf = UnionFind::new(variables);
    let mut map: HashMap<Variable, Type> = HashMap::new();
    let mut constraints = VecDeque::from(constraints);
    while !constraints.is_empty() {
        let constraint = constraints.pop_front().unwrap();
        match (constraint.type_l, constraint.type_r) {
            (Type::Bool, Type::Bool) | (Type::Num, Type::Num) | (Type::Unit, Type::Unit) => (),
            (Type::Var(l), Type::Var(r)) => {
                let l_root = uf.find(&l)?.clone();
                let r_root = uf.find(&r)?.clone();
                let root = uf.union(&l, &r).unwrap();
                match (map.remove(&l_root), map.remove(&r_root)) {
                    (Some(lval), Some(rval)) => {
                        constraints.push_back(Constraint {
                            type_l: lval.clone(),
                            type_r: rval,
                            expr_l: format!("val({})", l),
                            expr_r: format!("val({})", r),
                        });
                        Some(lval)
                    }
                    (Some(val), None) | (None, Some(val)) => Some(val),
                    (None, None) => None,
                }
                .and_then(|v| map.insert(root.clone(), v));
            }
            (Type::Var(var), val) | (val, Type::Var(var)) if !val.contains_var(&var) => {
                match map.insert(var, val.clone()) {
                    None => (),
                    Some(old_val) => {
                        constraints.push_back(Constraint {
                            type_l: old_val,
                            type_r: val,
                            expr_l: constraint.expr_l,
                            expr_r: constraint.expr_r,
                        });
                    }
                }
            }
            (
                Type::Fn {
                    arg: arg_l,
                    ret: ret_l,
                },
                Type::Fn {
                    arg: arg_r,
                    ret: ret_r,
                },
            ) => {
                constraints.extend(
                    vec![
                        Constraint {
                            type_l: *arg_l,
                            type_r: *arg_r,
                            expr_l: format!("arg({})", constraint.expr_l),
                            expr_r: format!("arg({})", constraint.expr_r),
                        },
                        Constraint {
                            type_l: *ret_l,
                            type_r: *ret_r,
                            expr_l: format!("ret({})", constraint.expr_l),
                            expr_r: format!("ret({})", constraint.expr_r),
                        },
                    ]
                    .into_iter(),
                );
            }
            (
                Type::Product {
                    left: l_1,
                    right: r_1,
                },
                Type::Product {
                    left: l_2,
                    right: r_2,
                },
            )
            | (
                Type::Sum {
                    left: l_1,
                    right: r_1,
                },
                Type::Sum {
                    left: l_2,
                    right: r_2,
                },
            ) => {
                constraints.extend(
                    vec![
                        Constraint {
                            type_l: *l_1,
                            type_r: *l_2,
                            expr_l: format!("({}).L", constraint.expr_l),
                            expr_r: format!("({}).L", constraint.expr_r),
                        },
                        Constraint {
                            type_l: *r_1,
                            type_r: *r_2,
                            expr_l: format!("({}).R", constraint.expr_l),
                            expr_r: format!("({}).R", constraint.expr_r),
                        },
                    ]
                    .into_iter(),
                );
            }
            (lhs, rhs) => return Err(format!("Unification failed with type {lhs} and {rhs}")),
        }
    }
    let updated_map = map
        .into_iter()
        .map(|(x, y)| (uf.find(&x).unwrap().clone(), y))
        .collect();
    Ok((uf, updated_map))
}

fn get_type(mut tau: Type, mut uf: UnionFind<Variable>, mut map: HashMap<Variable, Type>) -> Type {
    loop {
        let vars = tau.get_free_vars();
        if vars.is_empty() {
            return tau;
        }
        let x = vars.first().unwrap();
        let r_opt = uf.find(x);
        if let Ok(r) = r_opt
            && x != r
        {
            tau = tau.substitute(x.clone(), Type::Var(r.clone()));
            continue;
        }
        match r_opt.ok().and_then(|r| map.remove(r)) {
            Some(t_x) => tau = tau.substitute(x.clone(), t_x),
            None => tau = tau.add_one_quantifier(x.clone()),
        }
    }
}

File: src/TypeCheck.hs
module TypeCheck where

import AST
import ASTUtil
import Control.Monad (unless, when)
import Control.Monad.Trans.Class (lift)
import Control.Monad.Trans.State.Strict hiding (put)
import Data.List (nub)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Maybe (fromMaybe)
import Debug.Trace (trace)
import Flags (Mode (..))
import UnionFind

data Constraint = Constraint
  { typeL :: Type,
    typeR :: Type
  }
  deriving (Eq)

instance Show Constraint where
  show c = show (typeL c) ++ " = " ++ show (typeR c)

formatConstraints :: [Constraint] -> String
formatConstraints [] = "No constraints."
formatConstraints cs =
  "Constraints:\n"
    ++ unlines ["  " ++ show c | c <- cs]

type VarId = Int

type TCM a = StateT VarId (Either String) a

freshTypeVar :: TCM Type
freshTypeVar = do
  n <- get
  modify (+ 1)
  return $ TVar (VName ("type_" ++ show n))

instantiate :: Type -> TCM Type
instantiate t = case t of
  TNum -> return t
  TBool -> return t
  TUnit -> return t
  TVar _ -> return t
  TFn a r -> TFn <$> instantiate a <*> instantiate r
  TProduct l r -> TProduct <$> instantiate l <*> instantiate r
  TSum l r -> TSum <$> instantiate l <*> instantiate r
  TForall a tau -> do
    newVar <- freshTypeVar
    subst <- instantiate tau
    return (substituteMapVar (Map.singleton a newVar) subst)
  TMu a body -> TMu a <$> instantiate body

typeCheck :: Mode -> Expr -> Either String Type
typeCheck mode expr = do
  let result0 = runStateT (getConstraints expr []) 0
  case result0 of
    Left err -> Left err
    Right ((curType, constraints), _) -> do
      () <- if mode == VeryVerbose then trace (formatConstraints constraints) (Right ()) else Right ()
      uf <- unification constraints
      let result = getType curType uf
      unless (null (getFreeVars result)) $
        Left "Free variables remain in type"
      return result

getConstraints :: Expr -> [Type] -> TCM (Type, [Constraint])
getConstraints expr ctx = case expr of
  ENum _ -> return (TNum, [])
  ETrue -> return (TBool, [])
  EFalse -> return (TBool, [])
  EAddop _ left right -> do
    (tauLeft, cLeft) <- getConstraints left ctx
    (tauRight, cRight) <- getConstraints right ctx
    let constraints =
          cLeft
            ++ cRight
            ++ [ Constraint tauLeft TNum,
                 Constraint tauRight TNum
               ]
    return (TNum, constraints)
  EMulop _ left right -> do
    (tauLeft, cLeft) <- getConstraints left ctx
    (tauRight, cRight) <- getConstraints right ctx
    let constraints =
          cLeft
            ++ cRight
            ++ [ Constraint tauLeft TNum,
                 Constraint tauRight TNum
               ]
    return (TNum, constraints)
  ERelop _ left right -> do
    (tauLeft, cLeft) <- getConstraints left ctx
    (tauRight, cRight) <- getConstraints right ctx
    let constraints =
          cLeft
            ++ cRight
            ++ [ Constraint tauLeft TNum,
                 Constraint tauRight TNum
               ]
    return (TBool, constraints)
  EIf cond then_ else_ -> do
    (tauCond, cCond) <- getConstraints cond ctx
    (tauThen, cThen) <- getConstraints then_ ctx
    (tauElse, cElse) <- getConstraints else_ ctx
    let constraints =
          cCond
            ++ cThen
            ++ cElse
            ++ [ Constraint tauCond TBool,
                 Constraint tauThen tauElse
               ]
    return (tauThen, constraints)
  EAnd left right -> do
    (tauLeft, cLeft) <- getConstraints left ctx
    (tauRight, cRight) <- getConstraints right ctx
    let constraints =
          cLeft
            ++ cRight
            ++ [ Constraint tauLeft TBool,
                 Constraint tauRight TBool
               ]
    return (TBool, constraints)
  EOr left right -> do
    (tauLeft, cLeft) <- getConstraints left ctx
    (tauRight, cRight) <- getConstraints right ctx
    let constraints =
          cLeft
            ++ cRight
            ++ [ Constraint tauLeft TBool,
                 Constraint tauRight TBool
               ]
    return (TBool, constraints)
  EVar (VName x) -> lift $ Left $ "Free variable: " ++ show x
  EVar (VDeBruijn depth) -> do
    let ctxLen = length ctx
    if depth >= ctxLen
      then lift $ Left "DeBruijn index out of bounds"
      else do
        let ty = ctx !! (ctxLen - 1 - depth)
        instTy <- instantiate ty
        return (instTy, [])
  ELam _ mt e -> do
    tauX <- maybe freshTypeVar return mt
    let ctx' = ctx ++ [tauX]
    (tauRet, cRet) <- getConstraints e ctx'
    return (TFn tauX tauRet, cRet)
  EApp lam arg -> do
    (tauLam, cLam) <- getConstraints lam ctx
    (tauArg, cArg) <- getConstraints arg ctx
    tauRet <- freshTypeVar
    let constraints =
          cLam
            ++ cArg
            ++ [Constraint tauLam (TFn tauArg tauRet)]
    return (tauRet, constraints)
  EPair left right -> do
    (tauL, cL) <- getConstraints left ctx
    (tauR, cR) <- getConstraints right ctx
    return (TProduct tauL tauR, cL ++ cR)
  EProject e d -> do
    (tauE, cE) <- getConstraints e ctx
    tauL <- freshTypeVar
    tauR <- freshTypeVar
    let constraints =
          cE
            ++ [Constraint tauE (TProduct tauL tauR)]
    return (case d of L -> tauL; R -> tauR, constraints)
  EUnit -> return (TUnit, [])
  EInject e d mt -> do
    (tauE, cE) <- getConstraints e ctx
    case mt of
      Just (TSum tauL tauR) -> do
        let tauFull = case d of
              L -> TSum tauE tauR
              R -> TSum tauL tauE
        let extra = Constraint tauE (case d of L -> tauL; R -> tauR)
        return (tauFull, cE ++ [extra])
      Just _ -> lift $ Left "Injection annotation must be a sum type"
      Nothing -> do
        tauOther <- freshTypeVar
        let tauFull = case d of
              L -> TSum tauE tauOther
              R -> TSum tauOther tauE
        return (tauFull, cE)
  ECase e xleft eleft xright eright -> do
    (tauSum, cSum) <- getConstraints e ctx
    tauL <- freshTypeVar
    tauR <- freshTypeVar

    let ctxL = ctx ++ [tauL]
    (tauLAfter, cL) <- getConstraints eleft ctxL
    let ctxR = ctx ++ [tauR]
    (tauRAfter, cR) <- getConstraints eright ctxR

    let constraints =
          cSum
            ++ cL
            ++ cR
            ++ [ Constraint tauSum (TSum tauL tauR),
                 Constraint tauLAfter tauRAfter
               ]
    return (tauLAfter, constraints)
  EFix _ e -> do
    tauX <- freshTypeVar
    let ctx' = ctx ++ [tauX]
    (tauE, cE) <- getConstraints e ctx'
    let constraints = cE ++ [Constraint tauX tauE]
    return (tauX, constraints)
  ELet x mt e_x e_in -> do
    (tauX, cX) <- getConstraints e_x ctx
    (tauXGen, cX') <- case mt of
      Just (TForall v body) -> do
        newVar <- freshTypeVar
        let body' = substituteMapVar (Map.singleton v newVar) body
        let cX' = cX ++ [Constraint tauX body']
        tauXGen <- lift $ generalize tauX cX'
        return (tauXGen, cX')
      Just t -> lift $ do
        tauXGen <- generalize tauX (cX ++ [Constraint tauX t])
        return (tauXGen, cX ++ [Constraint tauX t])
      Nothing -> lift $ do
        tauXGen <- generalize tauX cX
        return (tauXGen, cX)
    let ctx' = ctx ++ [tauXGen]
    (tauIn, cIn) <- getConstraints e_in ctx'
    return (tauIn, cIn ++ cX')

generalize :: Type -> [Constraint] -> Either String Type
generalize tau constraints = do
  uf <- unification constraints
  let tauX = getType tau uf
  if not (null (getFreeVars tauX))
    then Left "Cannot generalize: free variables remain"
    else return tauX

unification :: [Constraint] -> Either String UnionFind
unification constraints = do
  let allVars = nub $ concatMap (\c -> getAllVars (typeL c) ++ getAllVars (typeR c)) constraints
  unification' mkUnionFind constraints

typeLevel :: Type -> Int
typeLevel (TVar _) = 0
typeLevel TNum = 2
typeLevel TBool = 2
typeLevel TUnit = 2
typeLevel _ = 1

unificationFailed :: Constraint -> Either String a
unificationFailed c = Left $ "Unification failed: " ++ show (typeL c) ++ " = " ++ show (typeR c)

unification' ::
  UnionFind ->
  [Constraint] ->
  Either String UnionFind
unification' uf [] = return uf
unification' uf (c : cs) = case (typeL c, typeR c) of
  (TNum, TNum) -> unification' uf cs
  (TBool, TBool) -> unification' uf cs
  (TUnit, TUnit) -> unification' uf cs
  (TVar l, TVar r) ->
    let rootL = find uf (TVar l)
        rootR = find uf (TVar r)
     in if rootL == rootR
          then unification' uf cs
          else case (rootL, rootR) of
            (TVar _, TVar _) ->
              let uf' = unionBy typeLevel uf (TVar l) (TVar r)
               in unification' uf' cs
            (TVar _, _) ->
              let uf' = uf {parent = Map.insert rootL rootR (parent uf)}
               in unification' uf' cs
            (_, TVar _) ->
              let uf' = uf {parent = Map.insert rootR rootL (parent uf)}
               in unification' uf' cs
            (TNum, TNum) -> unification' uf cs
            (TBool, TBool) -> unification' uf cs
            (TUnit, TUnit) -> unification' uf cs
            _ -> unification' uf (Constraint rootL rootR : cs)
  (TVar v, t) ->
    if not (containsVar v t)
      then
        let root = find uf (TVar v)
         in case root of
              TVar _ ->
                let uf' = unionBy typeLevel uf root t
                 in unification' uf' cs
              _ -> case (root, t) of
                (TNum, TNum) -> unification' uf cs
                (TBool, TBool) -> unification' uf cs
                (TUnit, TUnit) -> unification' uf cs
                (TNum, _) -> unificationFailed c
                (TBool, _) -> unificationFailed c
                (TUnit, _) -> unificationFailed c
                (_, TNum) -> unificationFailed c
                (_, TBool) -> unificationFailed c
                (_, TUnit) -> unificationFailed c
                _ -> unification' uf (Constraint root t : cs)
      else unificationFailed c
  (t, TVar v) -> unification' uf (Constraint (typeR c) (typeL c) : cs)
  (TFn a1 r1, TFn a2 r2) ->
    let newCs = [Constraint a1 a2, Constraint r1 r2]
     in unification' uf (newCs ++ cs)
  (TProduct l1 r1, TProduct l2 r2) ->
    let newCs = [Constraint l1 l2, Constraint r1 r2]
     in unification' uf (newCs ++ cs)
  (TSum l1 r1, TSum l2 r2) ->
    let newCs = [Constraint l1 l2, Constraint r1 r2]
     in unification' uf (newCs ++ cs)
  (TMu _ body, TMu _ body') ->
    if body == body'
      then unification' uf cs
      else unificationFailed c
  (TMu _ body, t) ->
    unification' uf (Constraint t (substituteVar (VDeBruijn 0) (typeL c) body) : cs)
  (t, TMu _ body) ->
    unification' uf (Constraint t (substituteVar (VDeBruijn 0) (typeR c) body) : cs)
  _ -> unificationFailed c

getType :: Type -> UnionFind -> Type
getType tau uf =
  case getFreeVars tau of
    [] -> tau
    (x : _) ->
      let r = find uf (TVar x)
       in if r == TVar x
            then getType (TForall x tau) uf
            else getType (substituteVar x r tau) uf

containsVar :: Variable -> Type -> Bool
containsVar v = go
  where
    go TNum = False
    go TBool = False
    go TUnit = False
    go (TVar v') = v == v'
    go (TFn a r) = go a || go r
    go (TProduct l r) = go l || go r
    go (TSum l r) = go l || go r
    go (TForall a tau') = a /= v && go tau'
    go (TMu a body) = a /= v && go body

substituteVar :: Variable -> Type -> Type -> Type
substituteVar v t = substituteMapVar (Map.singleton v t)

substituteMapVar :: Map Variable Type -> Type -> Type
substituteMapVar mp t = case t of
  TVar v -> fromMaybe t (Map.lookup v mp)
  TFn a r -> TFn (substituteMapVar mp a) (substituteMapVar mp r)
  TProduct l r -> TProduct (substituteMapVar mp l) (substituteMapVar mp r)
  TSum l r -> TSum (substituteMapVar mp l) (substituteMapVar mp r)
  TForall a tau' ->
    if a `Map.member` mp
      then substituteMapVar mp tau'
      else TForall a (substituteMapVar mp tau')
  TMu a body ->
    if a `Map.member` mp
      then substituteMapVar mp body
      else TMu a (substituteMapVar mp body)
  _ -> t